csrc/abstract_tensor.h

```diff
@@ -332,10 +851,131 @@ struct AbstractTensor {
   //   row2  id4    id5
   // in another word, the return value will be an AbstractTensor:
   // [dim0={id0, id1, id4}, dim1={id2, id3, id5}].
-  AbstractTensor& addRow(AbstractTensor tensor);
+  AbstractTensorWithInfo<Info>& addRow(AbstractTensorWithInfo<Info> tensor) {
+    NVF_CHECK(
+        domain.size() == tensor.domain.size(),
+        "Can not add a new row with different number of domains.");
+    NVF_CHECK(
+        std::all_of(
+            domain.begin(),
+            domain.end(),
+            [](const AbstractId& aid) { return aid.is<std::vector>(); }),
+        "Can not add a new row to an AbstractTensor with non-vector domains.")
+
+    for (auto i : c10::irange(domain.size())) {
+      domain[i].as<std::vector>().emplace_back(std::move(tensor.domain[i]));
+    }
+    return *this;
+  }
 
   // Remove all the null elements.
-  AbstractTensor& strip();
+  AbstractTensorWithInfo<Info>& strip() {
+    AbstractTensorWithInfo<Info> result;
+    for (const auto& aid : domain) {
+      if (aid.hasValue()) {
+        result.pushBack(aid);
+      }
+    }
+    std::swap(result, *this);
+    return *this;
+  }
+};
+
+struct EmptyInfo {
+  static EmptyInfo merge(const EmptyInfo&, const EmptyInfo&) {
+    return {};
+  }
+
+  bool operator==(const EmptyInfo& t) const {
+    return true;
+  }
+};
+
+using AbstractTensor = AbstractTensorWithInfo<EmptyInfo>;
+
+//! This is a holds a comparable type of "tag". This version is strict in the
+//! sense that we check that the tags match whenever we do a merge instead of
+//! allowing mismatches.
+template <typename Tag>
+struct TagSetInfo {
+  std::unordered_set<Tag> tags;
+  static TagSetInfo<Tag> merge(
+      const TagSetInfo<Tag>& a,
+      const TagSetInfo<Tag>& b) {
+    TagSetInfo<Tag> merged_tag_info{a.tags};
+    merged_tag_info.tags.insert(b.tags.begin(), b.tags.end());
+    return merged_tag_info;
+  }
+
+  bool operator==(TagSetInfo<Tag>&& t) const {
+    return tags == t.tags;
+  }
+};
+
+//! This is a special case of AbstractTensorWithInfo which propagates a set of
+//! tags for each axis. The tags can be any hashable type with equality;
+//! typically Tag would be an enum.
+template <typename Tag>
+struct TaggedAbstractTensor : AbstractTensorWithInfo<TagSetInfo<Tag>> {
+  TaggedAbstractTensor(
+      std::vector<AbstractId> domain,
+      std::vector<std::unordered_set<Tag>> tag_sets)
+      : AbstractTensorWithInfo<TagSetInfo<Tag>>(domain) {
+    NVF_ERROR(
+        this->info.size() == tag_sets.size(),
+        "Mismatch in length between provided domain and tag sets");
+    size_t i = 0;
+    for (const auto& s : tag_sets) {
+      this->info[i++].tags.insert(s.begin(), s.end());
+    }
+  }
```

Comment:
nit: If we add a new ctor `TagSetInfo(std::unordered_set<Tag>)`, then we can simplify this as:
```suggestion
  TaggedAbstractTensor(
      std::vector<AbstractId> domain,
      std::vector<std::unordered_set<Tag>> tag_sets)
      : AbstractTensorWithInfo<TagSetInfo<Tag>>(domain, {tag_sets.begin(), tag_sets.end()}) {}
```

Reply:
Great suggestion. That did clean up the `TaggedAbstractTensor` constructors.